### PR TITLE
Add scripts extract and transform the

### PR DESCRIPTION
**Pull request by GeekAngus**
## Types of changes

- **New feature**

## Description
Background: Start from 2022, we extract the KKTIX data via KKTIX API and load to "pycontw-225217.ods.ods_kktix_attendeeId_datetime". However most of the data are store in the ATTENDEE_INFO column with json format. To use metabase with SQL, users need to extract the data by json_extract with the knowledge kktix format instead of flat database. And we also need to rewrite all the SQLs build for current databases.
Solution: Transform the tables in backend that we could keep the same user experience by using Metabase.

## Checklist:
- [ ] Add test cases to all the changes you introduce
- [ ] Run `poetry run pytest` locally to ensure all linter checks pass
- [ ] Update the documentation if necessary

## Steps to Test This Pull Request
`./kktix_bq_etl.py 2023`

## Expected behavior
The data had been load to ods_kktix_ticket_${ticket_type}_attendees_test on bigquery
ticket_type = corporate individual reserved

### CODE REVIEW

File: contrib/README.md [hunk 1]
```diff
@@ -17,4 +17,11 @@
...
-    3. Then run `python upload-survey-cake-csv-to-bigquery.py --year=<20xx> -p`. `-p` stands for `production`
\ No newline at end of file
+    3. Then run `python upload-survey-cake-csv-to-bigquery.py --year=<20xx> -p`. `-p` stands for `production`
+
+## KKTIX BigQuery Transform
+1. Background: Start from 2022, we extract the KKTIX data via KKTIX API and load to "pycontw-225217.ods.ods_kktix_attendeeId_datetime". However most of the data are store in the ATTENDEE_INFO column with json format. To use metabase with SQL, users need to extract the data by json_extract with the knowledge kktix format instead of flat database. And we also need to rewrite all the SQLs build for current databases.
+2. Solution: Transform the tables in backend that we could keep the same user experience by using Metabase.
+3. Run: 
+ - `./kktix_bq_etl.py -t ods_kktix_ticket_reserved_attendees_test -k reserved -y 2023 --upload`
+ - for 3 tables: `./kktix_bq_etl.py 2023`
```
Comment (david30907d) [inline contrib/README.md:27]: thx for documenting these up 🙏

File: contrib/kktix_bq_etl.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    "know_financial_aid",
+    "have_you_ever_attended_pycon_tw",
+    "pynight_attending_or_not",
+    "how_did_you_know_pycon_tw",
+}
+
+CANONICAL_COLUMN_NAMES_2020_EXTRA_RESERVED: Set = set()
+
+
+CANONICAL_COLUMN_NAMES_2019_CORE = {
```
Comment (david30907d) [inline contrib/kktix_bq_etl.py:76]: seems to me that these `canonical_xxx` sets are unused?

File: contrib/kktix_bq_etl.py [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+CANONICAL_COLUMN_NAMES_2018_EXTRA_INDIVIDUAL: Set = set()
+
+CANONICAL_COLUMN_NAMES_2018_EXTRA_RESERVED = {
+    "invoiced_company_name",
+    "unified_business_no",
+}
+
+
+HEURISTIC_COMPATIBLE_MAPPING_TABLE = {
+    # from 2020 reformatted column names
```
Comment (david30907d) [inline contrib/kktix_bq_etl.py:144]: thx for writting these comments 🙏

File: contrib/kktix_bq_etl.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+    "how_did_you_find_out_pycon_tw_pycon_tw": "how_did_you_know_pycon_tw",
+    "have_you_ever_attended_pycon_tw_pycon_tw": "have_you_ever_attended_pycon_tw",
+    "privacy_policy_of_pycon_tw_2020": "privacy_policy_of_pycon_tw",
+    "privacy_policy_of_pycon_tw_2020_pycon_tw_2020_bitly3eipaut": "privacy_policy_of_pycon_tw",
+    "ive_already_read_and_i_accept_the_privacy_policy_of_pycontw_2020_pycon_tw_2020": "ive_already_read_and_i_accept_the_privacy_policy_of_pycon_tw",
+    "ive_already_read_and_i_accept_the_privacy_policy_of_pycon_tw_2020_pycon_tw_2020": "ive_already_read_and_i_accept_the_privacy_policy_of_pycon_tw",
+    "ive_already_read_and_i_accept_the_epidemic_prevention_of_pycon_tw_2020_pycon_tw_2020_covid19": "ive_already_read_and_i_accept_the_epidemic_prevention_of_pycon_tw",
+    "do_you_know_we_have_financial_aid_this_year": "know_financial_aid",
+    "contact_email": "email",
+    # from 2020 reformatted column names which made it duplicate
```
Comment (david30907d) [inline contrib/kktix_bq_etl.py:161]: ditto

File: contrib/kktix_bq_etl.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+        # print(value_list)
+        df_data = pd.DataFrame()
+        df_data[column_list] = [value_list]
+
+        df_dict = pd.concat([df_dict, df_data], axis=1)
+        df = pd.concat([df, df_dict], ignore_index=True)
+
+    # print(df.columns)
+    sanitized_df = sanitize_column_names(df)
+    hash_privacy_info(sanitized_df)
```
Comment (david30907d) [inline contrib/kktix_bq_etl.py:521]: thx for remembering hashing these sensitive columns. However, we've already encrypted data on `extract` side. so no need to do it again~
https://github.com/pycontw/pycon-etl/blob/810096e2b8bbd711bd8e98b99744cd16356bdf50/dags/ods/kktix_ticket_orders/udfs/kktix_api.py#L33

File: contrib/kktix_bq_etl.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        # print(transc_date)    
+        df_dict['paid_date'] = transc_date
+
+        # ticket_type value is always "qrcode" here, and conflict with old table, 
+        # kyc, id_number, slot value is always empty
+        # 'id','ticket_id','state','checkin_code', 'qrcode', updated_at, order_no no exist in old table
+        # data will be handle later
+        useless_columns = ['id','ticket_id','state','checkin_code', 'qrcode', 'currency', 'data', 'ticket_type', 'kyc', 'updated_at', 'id_number', 'slot', 'order_no']
+        df_dict = df_dict.drop(columns = useless_columns) 
+        df_dict = df_dict.rename(columns = {"reg_no": "registration_no", "ticket_name": "ticket_type", "is_paid": "payment_status"})
```
Comment (david30907d) [inline contrib/kktix_bq_etl.py:500]: what do you thiink if we keep all of the original data in the warehouse, and only deserialize/extract those "useful" columns from JSON to another DB layer?

File: contrib/kktix_bq_etl.py (hunk 2 above)
Comment (GeekAngus) [reply contrib/kktix_bq_etl.py:76]: yes, the main code is from upload-kktix-ticket-csv-to-bigquery.py and I jus keep them, the set was used for unit test.

File: contrib/kktix_bq_etl.py (hunk 5 above)
Comment (GeekAngus) [reply contrib/kktix_bq_etl.py:521]: yes, could be removed after confirmed.

File: contrib/kktix_bq_etl.py (hunk 6 above)
Comment (GeekAngus) [reply contrib/kktix_bq_etl.py:500]: Yes, we should keep all the of the original data that maybe useful someday.

**Comment by GeekAngus** [comment]
Use #123 instead

**Pull request closed without merging**